pkg/zkproofs/ciphertext_validity.go

```diff
@@ -0,0 +1,170 @@
+	// Validate input
+	if pedersenOpening == nil {
+		return nil, errors.New("invalid randomness factor")
+	}
+
+	if pubKey == nil {
+		return nil, errors.New("invalid public key")
+	}
+
+	if ciphertext == nil || ciphertext.C == nil || ciphertext.D == nil {
+		return nil, errors.New("invalid ciphertext")
+	}
+
+	eg := elgamal.NewTwistedElgamal()
+
+	H := eg.GetH()
+	G := eg.GetG()
+
+	ed25519 := curves.ED25519()
+	// Convert message to a scalar
+	messageValue := new(big.Int).SetUint64(message)
+	x, _ := ed25519.Scalar.SetBigInt(messageValue)
+
+	// Step 1: Generate random blinding factors for the proof
+	rBlind := ed25519.Scalar.Random(crand.Reader) // Blinding factor for random value r
+	xBlind := ed25519.Scalar.Random(crand.Reader) // Blinding factor for random value x
+
+	// Step 2: Create commitments
+	rBlindH := H.Mul(rBlind)            // rBlind * H
+	xBlindG := G.Mul(xBlind)            // xBlind * G
+	Commitment1 := rBlindH.Add(xBlindG) // Commitment1 = rBlind * H + xBlind * G
+
+	Commitment2 := pubKey.Mul(rBlind) // Commitment2 = rBlind * P
+
+	// Step 3: Generate a challenge using the Fiat-Shamir heuristic.
+	// The challenge is basically just a hash of all the provided values. This locks in the values and makes sure that
+	// the proof cannot be for some other set of values.
+	transcript := NewProofTranscript()
+	transcript.AppendMessage("C1", Commitment1.ToAffineCompressed())
+	transcript.AppendMessage("C2", Commitment2.ToAffineCompressed())
+	challenge := transcript.ChallengeScalar()
+
+	// Step 4: Generate responses
+	Response1 := challenge.MulAdd(*pedersenOpening, rBlind) // Response1 = rBlind + challenge * pedersenOpening
+	Response2 := challenge.MulAdd(x, xBlind)                // Response2 = xBlind + challenge * x
+
+	return &CiphertextValidityProof{
+		Commitment1: Commitment1,
+		Commitment2: Commitment2,
+		Response1:   Response1,
+		Response2:   Response2,
+	}, nil
+}
+
+// VerifyCiphertextValidity verifies that some ciphertext is properly encrypted by the given Public Key, with the help
+// of proof sent by the prover.
+// Parameters:
+// - proof: The proof to verify.
+// - pubKey: The public key used in the encryption.
+// - ciphertext: The ciphertext to prove the validity of.
+func VerifyCiphertextValidity(proof *CiphertextValidityProof, pubKey curves.Point, ciphertext *elgamal.Ciphertext) bool {
+	// Validate input
+	if proof == nil || proof.Commitment1 == nil || proof.Commitment2 == nil || proof.Response1 == nil ||
+		proof.Response2 == nil || pubKey == nil || ciphertext == nil || ciphertext.C == nil || ciphertext.D == nil {
+		return false
+	}
+
+	eg := elgamal.NewTwistedElgamal()
+	H := eg.GetH()
+	G := eg.GetG()
+
+	// Step 0: Recompute the challenge using the Fiat-Shamir heuristic.
+	transcript := NewProofTranscript()
+	transcript.AppendMessage("C1", proof.Commitment1.ToAffineCompressed())
+	transcript.AppendMessage("C2", proof.Commitment2.ToAffineCompressed())
```

Comment:
Think we want to add ct.C and ct.D to the challenge as well so the proof can only be used as for the specific ciphertext.

Reply:
Done